Updating JSON docs, adding `cache_strings` to `ConfigDict` (#9178)
pydantic/pydantic: docs/concepts/json.md

In the future, `jiter` is intended to enable support validation errors to include the location
in the original JSON input which contained the invalid value.

### Partial JSON Parsing

**Starting in v2.7.0**, Pydantic's [JSON parser](https://docs.rs/jiter/latest/jiter/) offers support for partial JSON parsing, which is exposed via [`pydantic_core.from_json`][pydantic_core.from_json]. Here's an example of this feature in action:

```py
from pydantic_core import from_json

partial_json_data = '["aa", "bb", "c'  # (1)!

try:
    result = from_json(partial_json_data, allow_partial=False)
except ValueError as e:
    print(e)  # (2)!
    #> EOF while parsing a string at line 1 column 15

result = from_json(partial_json_data, allow_partial=True)
print(result)  # (3)!
#> ['aa', 'bb']
```

1. The JSON list is incomplete - it's missing a closing `"]`
2. When `allow_partial` is set to `False` (the default), a parsing error occurs.
3. When `allow_partial` is set to `True`, part of the input is deserialized successfully.

This also works for deserializing partial dictionaries. For example:

```py
from pydantic_core import from_json

> partial_dog_json = '{"breed": "lab", "name": "fluffy", "friends": ["buddy", "spot", "rufus"], "age}'

Reviewer: ```suggestion
partial_dog_json = '{"breed": "lab", "name": "fluffy", "friends": ["buddy", "spot", "rufus"], "age'
```

dog_dict = from_json(partial_dog_json, allow_partial=True)
print(dog_dict)
#> {'breed': 'lab', 'name': 'fluffy', 'friends': ['buddy', 'spot', 'rufus']}
```

!!! tip "Validating LLM Output"
    This feature is particularly beneficial for validating LLM outputs.
    We've written some blog posts about this topic, which you can find [here](https://blog.pydantic.dev/blog/category/llms/).

In future versions of Pydantic, we expect to expand support for this feature through either Pydantic's other JSON validation functions
([`pydantic.main.BaseModel.model_validate_json`][pydantic.main.BaseModel.model_validate_json] and
[`pydantic.type_adapter.TypeAdapter.validate_json`][pydantic.type_adapter.TypeAdapter.validate_json]) or model configuration. Stay tuned 🚀!

For now, you can use [`pydantic_core.from_json`][pydantic_core.from_json] in combination with [`pydantic.main.BaseModel.model_validate`][pydantic.main.BaseModel.model_validate] to achieve the same result. Here's an example:

```py
from pydantic_core import from_json

from pydantic import BaseModel

class Dog(BaseModel):
    breed: str
    name: str
    friends: list

> partial_dog_json = '{"breed": "lab", "name": "fluffy", "friends": ["buddy", "spot", "rufus"], "age}'

Reviewer: ```suggestion
partial_dog_json = '{"breed": "lab", "name": "fluffy", "friends": ["buddy", "spot", "rufus"], "age'
```
same

dog = Dog.model_validate(from_json(partial_dog_json, allow_partial=True))
print(repr(dog))
#> Dog(breed='lab', name='fluffy', friends=['buddy', 'spot', 'rufus'])
```

!!! example
    We recommend the following settings to ensure reliable parsing of partial JSON data:

>     1. For this to work reliably, fields on the model should have default values or be optional.

Reviewer: ```suggestion
    1. For this to work reliably, fields on the model should have default values.
```
Being optional doesn't actually matter.

>     2. Set `on_error='omit'` in the core schema for types that can fail. This is the default behavior for Pydantic fields.

Reviewer: Although this would work, it would mean any error on that field would cause the default value to be used, so I'm not sure it's a good idea to suggest this.

        """
        try:
            return handler(v)
        except ValidationError as exc:
            # there might be other types of errors resulting from partial JSON parsing
            # that you allow here, feel free to customize as needed
            if all(e['type'] == 'missing' for e in exc.errors()):
                raise pydantic_core.PydanticUseDefault()
            else:
                raise

    class NestedModel(BaseModel):
        x: int
        y: str

    class MyModel(BaseModel):
        foo: str | None = None
        bar: Annotated[
            tuple[str, int] | None, WrapValidator(default_on_error)
        ] = None
        nested: Annotated[
            NestedModel | None, WrapValidator(default_on_error)
        ] = None

    m = MyModel.model_validate(
        pydantic_core.from_json('{"foo": "x", "bar": ["world",', allow_partial=True)
    )
    print(repr(m))
    #> MyModel(foo='x', bar=None, nested=None)

    m = MyModel.model_validate(
        pydantic_core.from_json(
            '{"foo": "x", "bar": ["world", 1], "nested": {"x":', allow_partial=True
        )
    )
    print(repr(m))
    #> MyModel(foo='x', bar=('world', 1), nested=None)
    ```

> ### Caching Strings

Reviewer: This looks great.

In the future, `jiter` is intended to enable support validation errors to include the location
in the original JSON input which contained the invalid value.

### Partial JSON Parsing

> Starting in v2.7.0 and above, Pydantic's [JSON parser](https://docs.rs/jiter/latest/jiter/) offers support for partial JSON parsing, which is exposed via [`pydantic_core.from_json`][pydantic_core.from_json]. Here's an example of this feature in action:

Reviewer: Might be worth making this into an admonition.

`<pedant>`"Starting in v2.7.0 and above" is tautologous, I think it should either be "Starting in v2.7.0" or "In v2.7.0 and above"`</pedant>`

```py
from pydantic_core import from_json

partial_json_data = '["aa", "bb", "c'  # (1)!

try:
    result = from_json(partial_json_data, allow_partial=False)

> except ValueError as ve:
>     print(ve)  # (2)!

Reviewer: ```suggestion
except ValueError as e:
    print(e)  # (2)!
```

    #> EOF while parsing a string at line 1 column 15

result = from_json(partial_json_data, allow_partial=True)
print(result)  # (3)!
#> ['aa', 'bb']
```

1. The JSON list is incomplete - it's missing a closing `"]`

> 2. When `allow_partial` is set to `False`, a parsing error occurs.

Reviewer: ```suggestion
2. When `allow_partial` is set to `False` (the default), a parsing error occurs.
```

3. When `allow_partial` is set to `True`, part of the input is deserialized successfully.

This also works for deserializing partial dictionaries. For example:

```py
from pydantic_core import from_json

partial_dog_json = '{"breed": "lab", "name": "fluffy", "friends": ["buddy", "spot", "rufus"], "age}'
dog_dict = from_json(partial_dog_json, allow_partial=True)
print(dog_dict)
#> {'breed': 'lab', 'name': 'fluffy', 'friends': ['buddy', 'spot', 'rufus']}

> ```

Reviewer: Probably worth adding an example of 

```py
MyModel.model_validate(pydantic_core.from_json(data, allow_partial=True))
```

Reviewer: We also need to explain that:
1. Every field on the model needs to have a default for this to work reliably
2. You need to set `on_error='omit'` in the core schema for types that can fail.

Author: Done, good idea, thanks!

!!! tip "Validating LLM Output"
    This feature is particularly beneficial for validating LLM outputs.
    We've written some blog posts about this topic, which you can find [here](#TODO: insert link filtered on LLMs).

In future versions of Pydantic, we expect to expand support for this feature through either Pydantic's other JSON validation functions
([`pydantic.main.BaseModel.model_validate_json`][pydantic.main.BaseModel.model_validate_json] and
[`pydantic.type_adapter.TypeAdapter.validate_json`][pydantic.type_adapter.TypeAdapter.validate_json]) or model configuration. Stay tuned 🚀!

### Caching Strings

> Starting in v2.7.0 and above, Pydantic's [JSON parser](https://docs.rs/jiter/latest/jiter/) offers support for string caching.

Reviewer: as above.

> Starting in v2.7.0 and above, Pydantic's [JSON parser](https://docs.rs/jiter/latest/jiter/) offers support for string caching.

Reviewer: ```suggestion
Starting in v2.7.0 and above, Pydantic's [JSON parser](https://docs.rs/jiter/latest/jiter/) offers support for configuring how Python strings are cached during JSON parsing and when Python strings are constructed from Rust strings during Python validation (e.g. after `strip_whitespace=True`).
```

The `cache_strings` setting is exposed via both [model config][pydantic.config.ConfigDict] and [`pydantic_core.from_json`][pydantic_core.from_json].

The `cache_strings` setting can take any of the following values:

* `True` or `'all'` (the default): cache all strings

> * `keys`: cache only dictionary keys

Reviewer: ```suggestion
* `keys`: cache only dictionary keys, this only applies when used with [`pydantic_core.from_json`][pydantic_core.from_json] or when parsing JSON using [`Json`][TODO link to JSON meta type]
```

* `False` or `none`: no caching

> Using the string caching feature results in performance improvements, but increases memory usage slightly.

Reviewer: we should mention that:

1. Strings are cached using a fully associative cache with size [16,384](https://github.com/pydantic/jiter/blob/5bbdcfd22882b7b286416b22f74abd549c7b2fd7/src/py_string_cache.rs#L113), with a link to that code in rust
2. We should say that only strings where `len(string) < 64` (or is it `<=`?) are cached
3. Explain that there is some overhead to looking up the cache, which is normally worth it to avoid constructing strings, but if you know there will be very few repeated strings in your data, you might get a performance boost by disabling it.

Author: Done, thanks!
